packages/multicast_dns/lib/src/resource_record.dart

```diff
@@ -0,0 +1,393 @@
+  )   : assert(name != null),
+        assert(ResourceRecordType.debugAssertValid(resourceRecordType));
+
+  /// An A (IPv4) query.
+  ResourceRecordQuery.a(
+    String name, {
+    bool isMulticast = true,
+  }) : this(
+          ResourceRecordType.a,
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// An AAAA (IPv6) query.
+  ResourceRecordQuery.aaaa(
+    String name, {
+    bool isMulticast = true,
+  }) : this(
+          ResourceRecordType.aaaa,
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// A PTR (Server pointer) query.
+  ResourceRecordQuery.ptr(
+    String name, {
+    bool isMulticast = true,
+  }) : this(
+          ResourceRecordType.ptr,
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// An SRV (Service) query.
+  ResourceRecordQuery.srv(
+    String name, {
+    bool isMulticast = true,
+  }) : this(
+          ResourceRecordType.srv,
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// A TXT (Text record) query.
+  ResourceRecordQuery.txt(
+    String name, {
+    bool isMulticast = true,
+  }) : this(
+          ResourceRecordType.txt,
+          name,
+          isMulticast ? QuestionType.multicast : QuestionType.unicast,
+        );
+
+  /// Tye type of resource record - one of [ResourceRecordType]'s values.
+  final int resourceRecordType;
+
+  /// The Fully Qualified Domain Name associated with the request.
+  final String name;
+
+  /// The [QuestionType], i.e. multicast or unicast.
+  final int questionType;
+
+  /// Convenience accessor to determine whether the question type is multicast.
+  bool get isMulticast => questionType == QuestionType.multicast;
+
+  /// Convenience accessor to determine whether the question type is unicast.
+  bool get isUnicast => questionType == QuestionType.unicast;
+
+  /// Encodes this query to the raw wire format.
+  List<int> encode() {
+    return encodeMDnsQuery(
+      name,
+      type: resourceRecordType,
+      multicast: isMulticast,
+    );
+  }
+
+  @override
+  int get hashCode =>
+      _hashValues(<int>[resourceRecordType, name.hashCode, questionType]);
+
+  @override
+  bool operator ==(Object other) {
+    return other is ResourceRecordQuery &&
+        other.resourceRecordType == resourceRecordType &&
+        other.name == name &&
+        other.questionType == questionType;
+  }
+
+  @override
+  String toString() =>
+      '$runtimeType{$name, type: ${ResourceRecordType.toDebugString(resourceRecordType)}, isMulticast: $isMulticast}';
+}
+
+/// Base implementation of DNS resource records (RRs).
+abstract class ResourceRecord {
+  /// Creates a new ResourceRecord.
+  const ResourceRecord(this.resourceRecordType, this.name, this.validUntil)
+      : assert(name != null);
+
+  /// The FQDN for this record.
+  final String name;
```

Comment:
```suggestion
  final String domainName;
```

Reply:
Domain name will be confusing here because it's not actually the domain name people will be looking for - that's the domainName on the ptr record.